components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt

```diff
@@ -0,0 +1,190 @@
+        holdingIdentity: HoldingIdentity
+    ): List<StateAndRef<ContractState>> {
+
+        validateParameters(request)
+
+        @Suppress("UNCHECKED_CAST")
+        val componentGroups = entitySandboxService.get(holdingIdentity.toCorda())
+            .getEntityManagerFactory()
+            .transaction { em ->
+                val query = em.createNativeQuery(
+                    "SELECT tc.transaction_id, tc.group_idx, tc.leaf_idx, tc.data FROM " +
+                            "$UTXO_TX_COMPONENT_TABLE AS tc " +
+                            "JOIN $UTXO_VISIBLE_TX_TABLE AS visible_states " +
+                                "ON visible_states.transaction_id = tc.transaction_id " +
+                            "$whereJson " +
+                            "AND tc.group_idx IN (:groupIndices) " +
+                            "AND visible_states.created <= :$TIMESTAMP_LIMIT_PARAM_NAME",
+                    Tuple::class.java
+                ).setParameter("groupIndices", listOf(
+                    UtxoComponentGroup.OUTPUTS.ordinal,
+                    UtxoComponentGroup.OUTPUTS_INFO.ordinal)
+                )
+
+                request.parameters.filter { it.value != null }.forEach { rec ->
+                    val bytes = rec.value.array()
+                    query.setParameter(rec.key, serializationService.deserialize(bytes))
+                }
+
+                query.firstResult = request.offset
+                query.maxResults = request.limit
+
+                query.resultList as List<Tuple>
+            }.map { t ->
+                ComponentLeafDto(
+                    t[0] as String, // transactionId
+                    (t[1] as Number).toInt(), // groupIndex
+                    (t[2] as Number).toInt(), // leafIndex
+                    t[3] as ByteArray // data
+                )
+            }.groupBy { it.groupIndex }
+
+        val outputInfos = componentGroups[UtxoComponentGroup.OUTPUTS_INFO.ordinal]
+            ?.associateBy { it.transactionId to it.leafIndex }
+            ?: emptyMap()
+
+        val utxoComponentDtos = componentGroups[UtxoComponentGroup.OUTPUTS.ordinal]?.map {
+            val info = outputInfos[it.transactionId to it.leafIndex]
+
+            requireNotNull(info) {
+                "Missing output info at index [${it.leafIndex}] for UTXO transaction with ID [${it.transactionId}]"
+            }
+            UtxoTransactionOutputDto(it.transactionId, it.leafIndex, info.data, it.data)
+        } ?: emptyList()
+
+        return utxoComponentDtos.map {
+            val info = serializationService.deserialize<UtxoOutputInfoComponent>(it.info)
+            val contractState = serializationService.deserialize<ContractState>(it.data)
+            StateAndRefImpl(
+                state = TransactionStateImpl(contractState, info.notaryName, info.notaryKey, info.getEncumbranceGroup()),
+                ref = StateRef(parseSecureHash(it.transactionId), it.leafIndex)
+            )
+        }
+    }
```

Comment:
Why do we need two mapping steps here? I think we could construct the StateAndRefs in one go.
```suggestion
        return componentGroups[UtxoComponentGroup.OUTPUTS.ordinal]?.map {
            val serializedInfo = outputInfos[it.transactionId to it.leafIndex]

            requireNotNull(serializedInfo) {
                "Missing output info at index [${it.leafIndex}] for UTXO transaction with ID [${it.transactionId}]"
            }
            val info = serizializationService.deserialize<UtxoOutputInfoComponent>(serializedInfo.data)
            val contractState = serializationService.deserialize<ContractState>(it.data)
            StateAndRefImpl(
                state = TransactionStateImpl(contractState, info.notaryName, info.notaryKey, info.getEncumbranceGroup()),
                ref = StateRef(parseSecureHash(it.transactionId), it.leafIndex)
            )
          } ?: emptyList()
    }
```

Reply:
good point!